### PR TITLE
Add Pubsub for Actors

### PR DESCRIPTION
**Pull request by DiegoAJR**
# Description

Support Pubsub for Actors with new API endpoint and modified runtime to invoke actors' method.
Added API enpoint to publish for an specific actor type and id.
Added into Actors runtime configuration field to declare subscriptions
Modified runtime to handle subscriptions to topics.
Created e2e test: Publishing for an actor and invoking the right method. 

## Issue reference

WIP #501

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [X] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_2735_
* [X] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_2735_
* [X] Provided sample for the feature / Created issue in the https://github.com/dapr/python-sdk/ repo: dapr/python-sdk#_438_

### CODE REVIEW

File: pkg/grpc/api.go [hunk 1]
```diff
@@ -365,6 +366,112 @@ func (a *api) CallActor(ctx context.Context, in *internalv1pb.InternalInvokeRequ
 	return resp.Proto(), nil
 }
 
+func (a *api) PublishActorEvent(ctx context.Context, in *runtimev1pb.PublishActorEventRequest) (*emptypb.Empty, error) {
+	if a.pubsubAdapter == nil {
```
Comment (berndverst) [inline pkg/grpc/api.go:370]: @halspang can you think of a way to avoid code duplication here?

```
publishEventInternal(ctx context.Context, in interface{})
{

    switch in.(type) {
        case *runtimev1pb.PublishActorEventRequest: {
          // Actor publish path
        }
        case *runtimev1pb.PublishEventRequest: {
          // Regular publish path
        }
    }
}
```

or some generic way?

Essentially we should have the same code somehow for everything other than the addition of cloudevent fields.

Comment (halspang) [reply pkg/grpc/api.go:370]: Given that everything is the same except for the two new fields, I'd think you could just use a single common method. In that method either choose to add the fields if they're defined or, just add them anyway as they'll be null in the message. 

That last approach does add two attributes to all requests though, which I'm not sure everyone would be happy about so I'd go with the first approach.

**Comment by berndverst** [comment]
@halspang can you run the tests again?

File: pkg/http/api.go [hunk 2]
```diff
@@ -369,6 +369,12 @@ func (a *api) constructActorEndpoints() []Endpoint {
 			Version: apiVersionV1,
 			Handler: a.onRenameActorReminder,
 		},
+		{
+			Methods: []string{fasthttp.MethodPost, fasthttp.MethodPut},
+			Route:   "actors/{actorType}/{actorId}/publish/{pubsubname}/{topic:*}",
+			Version: apiVersionV1,
```
Comment (berndverst) [inline pkg/http/api.go:375]: this needs to be Alpha1

**Comment by yaron2** [comment]
@berndverst please re-review.

File: dapr/proto/runtime/v1/dapr.proto [hunk 3]
```diff
@@ -53,6 +53,9 @@ service Dapr {
   // Publishes events to the specific topic.
   rpc PublishEvent(PublishEventRequest) returns (google.protobuf.Empty) {}
 
+  // PublishActorEventAlpha1 Publishes an event from an actor with a specific topic.
+  rpc PublishActorEventAlpha1(PublishActorEventRequest) returns (google.protobuf.Empty) {}
```
Comment (artursouza) [inline dapr/proto/runtime/v1/dapr.proto:57]: Please, move this close to the other Actor methods. On the client side, this belongs to the ActorProxy IMO.

File: pkg/grpc/api.go [hunk 4]
```diff
@@ -73,6 +73,7 @@ type API interface {
 
 	// Dapr Service methods
 	PublishEvent(ctx context.Context, in *runtimev1pb.PublishEventRequest) (*emptypb.Empty, error)
+	PublishActorEventAlpha1(ctx context.Context, in *runtimev1pb.PublishActorEventRequest) (*emptypb.Empty, error)
```
Comment (artursouza) [inline pkg/grpc/api.go:76]: nit: move close to the other Actor methods.

File: pkg/http/api.go [hunk 5]
```diff
@@ -1879,11 +1885,20 @@ func (a *api) onShutdown(reqCtx *fasthttp.RequestCtx) {
 }
 
 func (a *api) onPublish(reqCtx *fasthttp.RequestCtx) {
+	a.onPubSubPublish(reqCtx, "", "")
+}
+
+func (a *api) onActorPublishAlpha1(reqCtx *fasthttp.RequestCtx) {
```
Comment (artursouza) [inline pkg/http/api.go:1891]: nit: just onActorPublish is enough since internal method name can change freely.

File: pkg/http/api.go [hunk 6]
```diff
@@ -1879,11 +1885,20 @@ func (a *api) onShutdown(reqCtx *fasthttp.RequestCtx) {
...
+	actortype := reqCtx.UserValue(actorTypeParam).(string)
```
Comment (artursouza) [inline pkg/http/api.go:1892]: nit: Maybe we can keep the original method and check for the presence of actorTypeParam and actorIdParam.

File: pkg/runtime/pubsub/cloudevents.go [hunk 7]
```diff
@@ -29,13 +29,15 @@ type CloudEvent struct {
 	DataContentType string
 	TraceID         string
 	TraceState      string
+	ActorType       string
```
Comment (artursouza) [inline pkg/runtime/pubsub/cloudevents.go:32]: Do we want to rename it to "DaprActorType" and "DaprActorId" to avoid conflict with existing names applications might be using?

File: pkg/runtime/runtime.go [hunk 8]
```diff
@@ -573,14 +573,9 @@ func (a *DaprRuntime) beginPubSub(subscribeCtx context.Context, name string, ps
 	case GRPCProtocol:
 		publishFunc = a.publishMessageGRPC
 	}
-	topicRoutes, err := a.getTopicRoutes()
```
Comment (artursouza) [inline pkg/runtime/runtime.go:576]: Why did you remove the error handling here?

File: pkg/runtime/runtime.go [hunk 9]
```diff
@@ -695,6 +690,110 @@ func (a *DaprRuntime) beginPubSub(subscribeCtx context.Context, name string, ps
 		}
 	}
 
+	for _, actorConfigSubscriptions := range a.appConfig.Pubsub {
+		if actorConfigSubscriptions.PubSubName == name {
+			allowed := a.isPubSubOperationAllowed(name, actorConfigSubscriptions.Topic, a.scopedSubscriptions[name])
```
Comment (artursouza) [inline pkg/runtime/runtime.go:695]: I wonder if we can reuse the subscription code above. It is odd that we need to implement DLQ and TTL again here.

File: pkg/runtime/runtime.go [hunk 10]
```diff
@@ -695,6 +690,110 @@ func (a *DaprRuntime) beginPubSub(subscribeCtx context.Context, name string, ps
...
+	return nil
+}
+
+// actorCallPubsub calls the actor that was declared in the cloudevent or in the app.Config
+func (a *DaprRuntime) callActorPubsub(ctx context.Context, cloudEvent map[string]interface{}, actorPubSubSubscription config.PubSubConfig) error {
+	sendData, _ := json.Marshal(cloudEvent[pubsub.DataField])
+	if cloudEvent[pubsub.ActorIdField] == "" {
+		// Actor id is not in the cloudevent
+		if actorPubSubSubscription.ActorIDDataAttribute == "" {
+			log.Debug("Actor Pubsub: ActorID and ActorIDDataAttribute not declared in cloud event id - %s", cloudEvent[pubsub.IDField])
```
Comment (artursouza) [inline pkg/runtime/runtime.go:755]: Log Warning here. Users probably want to know the event was ignored.

Should these go to the DLQ instead?

File: pkg/runtime/runtime.go [hunk 11]
```diff
@@ -695,6 +690,110 @@ func (a *DaprRuntime) beginPubSub(subscribeCtx context.Context, name string, ps
...
+			if attribute, ok := dataMap[actorPubSubSubscription.ActorIDDataAttribute]; ok {
+				req := invokev1.NewInvokeMethodRequest(actorPubSubSubscription.Method)
+				req.WithActor(actorPubSubSubscription.ActorType, attribute)
+				req.WithHTTPExtension(nethttp.MethodPost, "")
+				req.WithRawData(sendData, contenttype.CloudEventContentType)
+				log.Infof("Invoking Actor %+v: with method %s", req.Actor(), req.Message().GetMethod())
+
+				res, err := a.actor.Call(ctx, req)
+				if err != nil {
+					log.Debugf("Error invoking actor: %s", err.Error())
```
Comment (artursouza) [inline pkg/runtime/runtime.go:773]: We probably want an log error here?

File: pkg/runtime/runtime.go [hunk 12]
```diff
@@ -695,6 +690,110 @@ func (a *DaprRuntime) beginPubSub(subscribeCtx context.Context, name string, ps
...
+					return err
+				}
+				log.Debugf("Actor response: %v", res.Message().GetData())
+			} else {
+				log.Debug("Actor Pubsub: Not a valid ActorIDAttribute for this actor")
```
Comment (artursouza) [inline pkg/runtime/runtime.go:778]: Log warning? DLQ?

File: pkg/runtime/runtime.go [hunk 13]
```diff
@@ -695,6 +690,110 @@ func (a *DaprRuntime) beginPubSub(subscribeCtx context.Context, name string, ps
...
+			}
+		}
+	} else {
+		// actor id is in the cloud event
+		actorid := fmt.Sprint(cloudEvent[pubsub.ActorIdField])
+		req := invokev1.NewInvokeMethodRequest(actorPubSubSubscription.Method)
```
Comment (artursouza) [inline pkg/runtime/runtime.go:784]: This code block seems to be very similar to the invocation above. Can this be refactored so we reuse the actor invocation logic?

File: tests/e2e/actor_pubsub/actor_pubsub_test.go [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+	"github.com/stretchr/testify/require"
+	"go.uber.org/ratelimit"
+)
+
+var tr *runner.TestRunner
+
+const (
+	numberOfMessagesToPublish = 60
+	publishRateLimitRPS       = 25
+	defaultActorType          = "myActorType"
```
Comment (artursouza) [inline tests/e2e/actor_pubsub/actor_pubsub_test.go:43]: Please, use an actor type that is specific to this test to avoid conflicting with other test apps.

File: tests/perf/actor_pubsub/actor_pubsub_test.go [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	numHealthChecks        = 60 // Number of times to check for endpoint health per app.
+	serviceApplicationName = "perf-actor-subscriber"
+	clientApplicationName  = "perf-actor-publisher"
+	actorType              = "testactorpubsub"
+	defaultpubsubName      = "messagebus"
+	defaultTopicNameHTTP   = "mytopic"
```
Comment (artursouza) [inline tests/perf/actor_pubsub/actor_pubsub_test.go:39]: pick something more unique.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: pkg/runtime/pubsub/cloudevents.go (hunk 7 above)
Comment (DiegoAJR) [reply pkg/runtime/pubsub/cloudevents.go:32]: We could rename them. Will require update in the components-contrib/pubsub/envelope.go file

File: pkg/grpc/api.go [hunk 16]
```diff
@@ -412,21 +443,23 @@ func (a *api) PublishEvent(ctx context.Context, in *runtimev1pb.PublishEventRequ
...
-			DataContentType: in.DataContentType,
+			Topic:           in.topic,
+			DataContentType: in.datacontectType,
 			Data:            body,
 			TraceID:         corID,
 			TraceState:      traceState,
-			Pubsub:          in.PubsubName,
+			Pubsub:          in.pubsubName,
+			ActorType:       actorType,
+			ActorID:         actorID,
```
[the remark this answers is not in the document]
Comment (DiegoAJR) [reply pkg/grpc/api.go:451]: We could declare a different cloud event structure. Update in components-contrib

**Comment by andrewhathaway** [comment]
@DiegoAJR @halspang @artursouza Hi all, appreciate this PR has not changed in the last 67 days however this is still a feature we would still greatly benefit from and are having to work around it in the meantime. Is this functionality still something we can expect to be added in the future? :)

**Comment by rotcafer** [comment]
@andrewhathaway can I ask how you are working around this now?

**Comment by paule96** [comment]
@DiegoAJR, @andrewhathaway, @rotcafer, @halspang, @artursouza 

would it make sense to take these changes and continue working on it, on an updatodate branch?
I could maybe invest my next week to work on this. But I'm not a Go programmer so probably my progress wouldn't be that fast.

But having pub/sub in actors would be so nice. As far as I understand this PR was almost done, just some minor adjustments were missing.

What are your opinions on it?

**Comment by artursouza** [comment]
This PR should not be merged. First, we should have a proposal to review some design decisions. If you want to take action in creating the proposal, see the Dapr/proposals repo.
